sdk/core/perf/src/program.cpp

```diff
@@ -278,152 +288,172 @@ void Azure::Perf::Program::Run(
-    // Wait for all tests to complete setUp
-    for (auto& t : tasks)
+
+    /******************** WarmUp ******************************/
+    if (options.Warmup)
     {
-      t.join();
+      RunTests(context, parallelTest, options, "Warmup", true);
     }
-  }
-
-  /******************** WarmUp ******************************/
-  if (options.Warmup)
-  {
-    RunTests(context, parallelTest, options, "Warmup", true);
-  }
 
-  /******************** Tests ******************************/
-  std::string iterationInfo;
-  try
-  {
-    for (int iteration = 0; iteration < options.Iterations; iteration++)
+    /******************** Tests ******************************/
+    std::string iterationInfo;
+    try
     {
-      if (iteration > 0)
+      for (int iteration = 0; iteration < options.Iterations; iteration++)
       {
-        iterationInfo.append(FormatNumber(iteration));
+        if (iteration > 0)
+        {
+          iterationInfo.append(FormatNumber(iteration));
+        }
+        RunTests(context, parallelTest, options, "Test" + iterationInfo);
       }
-      RunTests(context, parallelTest, options, "Test" + iterationInfo);
     }
-  }
-  catch (std::exception const& error)
-  {
-    std::cout << "Error: " << error.what();
-  }
-
-  std::cout << std::endl << "=== Pre-Cleanup ===" << std::endl;
-  {
-    if (!options.TestProxies.empty())
+    catch (std::exception const& error)
     {
-      std::cout << " - Deleting test recordings from test-proxies..." << std::endl;
-      std::cout << " - Disabling test-proxy playback" << std::endl;
+      std::cout << "Error: " << error.what();
     }
 
-    std::vector<std::thread> tasks(parallelTasks);
-    for (int i = 0; i < parallelTasks; i++)
+    std::cout << std::endl << "=== Pre-Cleanup ===" << std::endl;
     {
-      tasks[i] = std::thread([&parallelTest, i]() { parallelTest[i]->PreCleanUp(); });
+      if (!options.TestProxies.empty())
+      {
+        std::cout << " - Deleting test recordings from test-proxies..." << std::endl;
+        std::cout << " - Disabling test-proxy playback" << std::endl;
+      }
+
+      std::vector<std::thread> tasks(parallelTasks);
+      for (int i = 0; i < parallelTasks; i++)
+      {
+        tasks[i] = std::thread([&parallelTest, i]() { parallelTest[i]->PreCleanUp(); });
+      }
+      // Wait for all tests to complete setUp
+      for (auto& t : tasks)
+      {
+        t.join();
+      }
     }
-    // Wait for all tests to complete setUp
-    for (auto& t : tasks)
+
+    /******************** Clean up ******************************/
+    if (!options.NoCleanup)
     {
-      t.join();
+      std::cout << std::endl << "=== Cleanup ===" << std::endl;
+      std::vector<std::thread> tasks(parallelTasks);
+      for (int i = 0; i < parallelTasks; i++)
+      {
+        tasks[i] = std::thread([&parallelTest, i]() { parallelTest[i]->Cleanup(); });
+      }
+      for (auto& t : tasks)
+      {
+        t.join();
+      }
+      test->GlobalCleanup();
```

Comment:
If we take an exception this won't get run.  Is that intended or should this be in a finally?

Reply:
The process is in the process of shutting down because of the exception. So it's not going to be run either way.  And Mike indicated that the correct behavior in the presence of an exception is to abort the process.